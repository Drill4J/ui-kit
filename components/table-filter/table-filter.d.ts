interface Props {
    className?: string;
    filters: string[];
}
export declare const TableFilter: ({ className, filters }: Props) => JSX.Element;
export {};
