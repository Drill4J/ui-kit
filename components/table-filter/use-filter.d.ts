/// <reference types="react" />
interface FilterItem {
    name: string;
    checked: boolean;
}
declare type FilterState = {
    filterItems: FilterItem[];
    isExpanded: boolean;
};
declare type Action = ReturnType<typeof toggleFilter | typeof toggleAll | typeof setIsExpanded>;
export declare const toggleFilter: (name: string) => {
    readonly type: "TOGGLE_FILTER";
    readonly payload: string;
};
export declare const toggleAll: (checked: boolean) => {
    readonly type: "TOGGLE_ALL";
    readonly payload: boolean;
};
export declare const setIsExpanded: (checked: boolean) => {
    readonly type: "SET_IS_EXPANDED";
    readonly payload: boolean;
};
export declare const reducer: ({ filterItems, isExpanded }: FilterState, action: Action) => FilterState;
export declare const useFilter: (filters: string[]) => {
    isExpanded: boolean;
    filterItems: FilterItem[];
    selectedFilterItems: FilterItem[];
    dispatch: import("react").Dispatch<{
        readonly type: "TOGGLE_FILTER";
        readonly payload: string;
    } | {
        readonly type: "TOGGLE_ALL";
        readonly payload: boolean;
    } | {
        readonly type: "SET_IS_EXPANDED";
        readonly payload: boolean;
    }>;
};
export {};
