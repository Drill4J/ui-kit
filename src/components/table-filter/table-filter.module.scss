@import '../../style-guide/index';

.table-filter {
  position: relative;
  cursor: pointer;

  &__selected-items {
    font-family: $semibold-font;
    font-size: 12px;
    color: map-get($primary-blue, default);

    &:not(:last-child)::after {
      content: ', ';
    }
  }

  &__item-list-wrapper {
    position: absolute;
    right: 50%;
    top: 50%;
    transform: translate(65%, 10%);
    z-index: 150;
  }

  &__items-list {
    display: flex;
    flex-direction: column;
    min-width: 100px;
    max-height: 230px;
    overflow-y: auto;
    background-color: map-get($monochrome, white);
    box-shadow: 0 0 24px 0 rgba(0, 0, 0, 0.15);
  }

  &__item {
    display: flex;
    flex-direction: row;
    align-items: center;
    min-height: 32px;
    padding: 0px 8px;
    font-size: 12px;
    color: map-get($monochrome, black);
    cursor: pointer;

    &:hover {
      background-color: map-get($monochrome, light-tint);
    }
  }
}
