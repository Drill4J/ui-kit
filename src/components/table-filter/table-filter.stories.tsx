import { storiesOf } from '@storybook/react';

import { TableFilter } from './table-filter';

storiesOf('TableFilter', module).add('Filter with checkboxes',
  () => (
    <div style={{ display: 'flex', justifyContent: 'center' }}>
      <TableFilter filters={['New', 'Deleted', 'Mod', 'Unaffected', '5', '6', '7', '8', 'more8']} />
    </div>
  ));
