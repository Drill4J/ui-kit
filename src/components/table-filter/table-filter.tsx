import * as React from 'react';
import { BEM } from '@redneckz/react-bem-helper';
import { nanoid } from 'nanoid';

import { Checkbox } from '../../forms/inputs/checkbox';
import { useClickOutside } from '../../hooks';
import {
  useFilter, toggleFilter, toggleAll, setIsExpanded,
} from './use-filter';

import styles from './table-filter.module.scss';

interface Props {
  className?: string;
  filters: string[];
}

const tableFilter = BEM(styles);

export const TableFilter = tableFilter(({ className, filters }: Props) => {
  const {
    isExpanded, filterItems, selectedFilterItems, dispatch,
  } = useFilter(filters);
  const node = useClickOutside(() => dispatch(setIsExpanded(false)));

  return (
    <div className={className} ref={node}>
      <div onClick={() => dispatch(setIsExpanded(!isExpanded))}>
        {selectedFilterItems.length > 0 && selectedFilterItems.length < filters.length ? selectedFilterItems.map(({ name }) => (
          <SelectedItems key={name}>
            {name}
          </SelectedItems>
        )) : <SelectedItems>All</SelectedItems>}
      </div>
      {isExpanded && (
        <ItemsListWrapper>
          <ItemsList>
            <Item key={nanoid()}>
              <Checkbox
                label="All"
                checked={selectedFilterItems.length === filters.length}
                onChange={({ target: { checked } }) => dispatch(toggleAll(checked))}
              />
            </Item>
            {filterItems.map(({ name, checked }) => (
              <Item key={nanoid()}>
                <Checkbox
                  checked={checked}
                  label={name}
                  onChange={() => dispatch(toggleFilter(name))}
                />
              </Item>
            ))}
          </ItemsList>
        </ItemsListWrapper>
      )}
    </div>
  );
});

const SelectedItems = tableFilter.selectedItems('span');
const ItemsListWrapper = tableFilter.itemListWrapper('div');
const ItemsList = tableFilter.itemsList('div');
const Item = tableFilter.item('div');
