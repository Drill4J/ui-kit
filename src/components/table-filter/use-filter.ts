import { useEffect, useReducer } from 'react';

interface FilterItem {
  name: string;
  checked: boolean;
}

type FilterState = { filterItems: FilterItem[]; isExpanded: boolean };
type Action = ReturnType<typeof toggleFilter | typeof toggleAll | typeof setIsExpanded>;

const TOGGLE_FILTER = 'TOGGLE_FILTER';
const TOGGLE_ALL = 'TOGGLE_ALL';
const SET_IS_EXPANDED = 'SET_IS_EXPANDED';

export const toggleFilter = (name: string) => ({ type: TOGGLE_FILTER, payload: name } as const);
export const toggleAll = (checked: boolean) => ({ type: TOGGLE_ALL, payload: checked } as const);
export const setIsExpanded = (checked: boolean) => ({ type: SET_IS_EXPANDED, payload: checked } as const);

export const reducer = ({ filterItems, isExpanded }: FilterState, action: Action): FilterState => {
  switch (action.type) {
    case TOGGLE_FILTER:
      return {
        filterItems: filterItems.map(checkbox => {
          if (checkbox.name === action.payload) {
            return ({ ...checkbox, checked: !checkbox.checked });
          }
          return checkbox;
        }),
        isExpanded,
      };
    case TOGGLE_ALL:
      return { filterItems: filterItems.map((checkbox) => ({ ...checkbox, checked: action.payload })), isExpanded };
    case SET_IS_EXPANDED:
      return { filterItems, isExpanded: action.payload };
    default:
      return { filterItems, isExpanded };
  }
};

export const useFilter = (filters: string[]) => {
  const [{ filterItems, isExpanded }, dispatch] = useReducer(
    reducer,
    { filterItems: filters.map(filter => ({ name: filter, checked: true })), isExpanded: false },
  );
  const selectedFilterItems = filterItems.filter((selectedFilterItem) => selectedFilterItem.checked);

  useEffect(() => {
    filterItems.every(checkbox => checkbox.checked === false) && dispatch(toggleAll(true));
  }, [isExpanded]);

  return {
    isExpanded, filterItems, selectedFilterItems, dispatch,
  };
};
