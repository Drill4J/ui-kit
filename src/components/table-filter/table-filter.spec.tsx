import React from 'react';
import renderer from 'react-test-renderer';

import { TableFilter } from './table-filter';

describe('TableFilter', () => {
  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <TableFilter
          filters={['New, Mod, Deleted']}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
