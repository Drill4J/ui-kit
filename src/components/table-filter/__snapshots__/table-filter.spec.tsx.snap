// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableFilter should match snapshot 1`] = `
<div
  className="table-filter"
>
  <div
    onClick={[Function]}
  >
    <span
      className="table-filter__selected-items"
      data-modifiers=""
    >
      All
    </span>
  </div>
</div>
`;
